File: languages/swift/XANDEncrypt.swift
import Foundation
import CryptoKit
import CommonCrypto

public class XANDEncrypt {
    private let basePassword: String
    private let keySize: Int = 32 // 256 bits
    private let nonceSize: Int = 16
    private let paddingSize: Int = 16
    
    public init(password: String) {
        self.basePassword = password
    }
    
    /// Convert password to SHA-256 hash using Swift CryptoKit
    /// CryptoKit provides optimized cryptographic operations with native performance
    /// Data type provides efficient binary data handling with automatic memory management
    private func hashPassword(_ password: String) -> Data {
        let data = password.data(using: .utf8)!
        return Data(SHA256.hash(data: data))
    }
    
    /// Generate HMAC-SHA256 using CommonCrypto framework
    /// CommonCrypto provides low-level cryptographic operations for HMAC
    /// Array conversion enables direct byte manipulation for C interop
    private func generateHMAC(key: Data, data: Data) -> Data {
        let keyBytes = [UInt8](key)
        let dataBytes = [UInt8](data)
        var hmac = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
        
        CCHmac(CCHmacAlgorithm(kCCHmacAlgSHA256), keyBytes, keyBytes.count, dataBytes, dataBytes.count, &hmac)
        return Data(hmac)
    }
    
    /// Generate three time-dependent keys using HMAC-SHA256
    /// withUnsafeBytes provides direct memory access for byte conversion
    /// bigEndian ensures consistent endianness across different platforms
    private func generateTimeShiftedKeys(baseKey: Data, timestamp: UInt64) -> (key1: Data, key2: Data, key3: Data, timestamp: UInt64) {
        let timeBuffer = withUnsafeBytes(of: timestamp.bigEndian) { Data($0) }
        
        // Create three different keys for the three layers
        let key1 = generateHMAC(key: baseKey, data: timeBuffer)
        let key2 = generateHMAC(key: key1, data: "layer2".data(using: .utf8)!)
        let key3 = generateHMAC(key: key2, data: "layer3".data(using: .utf8)!)
        
        return (key1: key1, key2: key2, key3: key3, timestamp: timestamp)
    }
    
    /// Generate cryptographically secure random nonce and padding
    /// SecRandomCopyBytes uses iOS/macOS secure random number generator
    /// withUnsafeMutableBytes provides direct memory access for random byte generation
    private func generateNonceAndPadding() -> (nonce: Data, padding: Data) {
        var nonce = Data(count: nonceSize)
        _ = nonce.withUnsafeMutableBytes { pointer in
            SecRandomCopyBytes(kSecRandomDefault, nonceSize, pointer.baseAddress!)
        }
        
        let paddingLength = Int(nonce[0]) % paddingSize + 1
        var padding = Data(count: paddingLength)
        _ = padding.withUnsafeMutableBytes { pointer in
            SecRandomCopyBytes(kSecRandomDefault, paddingLength, pointer.baseAddress!)
        }
        
        return (nonce: nonce, padding: padding)
    }
    
    /// First encryption layer: XOR and bit rotation operations
    /// Swift UInt8 provides 8-bit unsigned integers for byte-level operations
    /// Data.append() enables efficient byte-by-byte construction of result
    private func layer1Scramble(data: Data, key: Data) -> Data {
        var result = Data()
        let dataBytes = [UInt8](data)
        let keyBytes = [UInt8](key)
        
        for i in 0..<dataBytes.count {
            let keyByte = keyBytes[i % keyBytes.count]
            let dataByte = dataBytes[i]
            
            // Complex bitwise operations
            var scrambled = dataByte ^ keyByte
            scrambled = ((scrambled << 3) | (scrambled >> 5)) & 0xFF // Rotate left by 3
            scrambled = scrambled ^ (((keyByte << 1) | (keyByte >> 7)) & 0xFF)
            
            result.append(scrambled)
        }
        
        return result
    }
    
    /// Second encryption layer: Position-dependent bit rotation
    /// Rotation amount calculated from byte position and key value
    /// Conditional rotation prevents unnecessary operations when rotation is zero
    private func layer2Manipulate(data: Data, key: Data) -> Data {
        var result = Data()
        let dataBytes = [UInt8](data)
        let keyBytes = [UInt8](key)
        
        for i in 0..<dataBytes.count {
            let keyByte = keyBytes[i % keyBytes.count]
            let dataByte = dataBytes[i]
            
            // Position-based bit rotation
            let rotation = (i + Int(keyByte)) % 8
            var manipulated = dataByte
            
            if rotation > 0 {
                manipulated = ((manipulated << rotation) | (manipulated >> (8 - rotation))) & 0xFF
            }
            
            // XOR with position-dependent key
            manipulated = manipulated ^ ((keyByte + UInt8(i)) & 0xFF)
            
            result.append(manipulated)
        }
        
        return result
    }
    
    /// Third encryption layer: AES-256-GCM encryption
    /// Uses Swift CryptoKit AES.GCM for authenticated encryption
    /// GCM mode provides both confidentiality and authenticity guarantees
    private func layer3Encrypt(data: Data, key: Data) throws -> (encrypted: Data, iv: Data) {
        var iv = Data(count: 16)
        _ = iv.withUnsafeMutableBytes { pointer in
            SecRandomCopyBytes(kSecRandomDefault, 16, pointer.baseAddress!)
        }
        
        let encrypted = try AES.GCM.seal(data, using: SymmetricKey(data: key), nonce: AES.GCM.Nonce(data: iv))
        return (encrypted: encrypted.combined!, iv: iv)
    }
    
    /// Layer 3: AES-256 decryption
    private func layer3Decrypt(data: Data, key: Data, iv: Data) throws -> Data {
        let sealedBox = try AES.GCM.SealedBox(combined: data)
        return try AES.GCM.open(sealedBox, using: SymmetricKey(data: key))
    }
    
    /// Apply key decay by XORing with time-based hash
    /// Prevents key reuse attacks by modifying keys after each operation
    /// Uses current timestamp to ensure uniqueness of decay value
    private func applyKeyDecay(keys: inout (key1: Data, key2: Data, key3: Data, timestamp: UInt64)) {
        let timestamp = UInt64(Date().timeIntervalSince1970 * 1000)
        
        // Apply decay to key1
        let decayValue1 = hashPassword("\(keys.key1.base64EncodedString())\(timestamp)")
        var decayedKey1 = Data()
        for i in 0..<keys.key1.count {
            decayedKey1.append(keys.key1[i] ^ decayValue1[i % decayValue1.count])
        }
        keys.key1 = decayedKey1
        
        // Apply decay to key2
        let decayValue2 = hashPassword("\(keys.key2.base64EncodedString())\(timestamp)")
        var decayedKey2 = Data()
        for i in 0..<keys.key2.count {
            decayedKey2.append(keys.key2[i] ^ decayValue2[i % decayValue2.count])
        }
        keys.key2 = decayedKey2
        
        // Apply decay to key3
        let decayValue3 = hashPassword("\(keys.key3.base64EncodedString())\(timestamp)")
        var decayedKey3 = Data()
        for i in 0..<keys.key3.count {
            decayedKey3.append(keys.key3[i] ^ decayValue3[i % decayValue3.count])
        }
        keys.key3 = decayedKey3
    }
    
    /// Main encryption function implementing triple-layer encryption
    /// Processes data through bit scrambling, position-dependent rotation, and AES-256
    /// Returns JSON string containing encrypted data and metadata
    public func encrypt(_ data: String) throws -> String {
        // Generate base key from password
        let baseKey = hashPassword(basePassword)
        
        // Generate time-shifted keys
        let timestamp = UInt64(Date().timeIntervalSince1970 * 1000)
        var keys = generateTimeShiftedKeys(baseKey: baseKey, timestamp: timestamp)
        
        // Generate nonce and padding
        let (nonce, padding) = generateNonceAndPadding()
        
        // Add padding to data
        var paddedData = data.data(using: .utf8)!
        paddedData.append(padding)
        
        // Layer 1: Bitwise scrambling
        let layer1Result = layer1Scramble(data: paddedData, key: keys.key1)
        
        // Layer 2: Advanced bit manipulation
        let layer2Result = layer2Manipulate(data: layer1Result, key: keys.key2)
        
        // Layer 3: AES encryption
        let layer3Result = try layer3Encrypt(data: layer2Result, key: keys.key3)
        
        // Apply key decay
        applyKeyDecay(keys: &keys)
        
        // Combine all components
        let result: [String: Any] = [
            "encrypted": layer3Result.encrypted.base64EncodedString(),
            "iv": layer3Result.iv.base64EncodedString(),
            "nonce": nonce.base64EncodedString(),
            "timestamp": timestamp,
            "padding_length": padding.count
        ]
        
        let jsonData = try JSONSerialization.data(withJSONObject: result)
        return String(data: jsonData, encoding: .utf8)!
    }
    
    /// Decrypt data using triple-layer decryption
    public func decrypt(_ encryptedData: String) throws -> String {
        let jsonData = encryptedData.data(using: .utf8)!
        let data = try JSONSerialization.jsonObject(with: jsonData) as! [String: Any]
        
        // Parse components
        let encrypted = Data(base64Encoded: data["encrypted"] as! String)!
        let iv = Data(base64Encoded: data["iv"] as! String)!
        let _nonce = Data(base64Encoded: data["nonce"] as! String)!
        let timestamp = data["timestamp"] as! UInt64
        let paddingLength = data["padding_length"] as! Int
        
        // Generate base key from password
        let baseKey = hashPassword(basePassword)
        
        // Regenerate time-shifted keys (must use same timestamp)
        let keys = generateTimeShiftedKeys(baseKey: baseKey, timestamp: timestamp)
        
        // Layer 3: AES decryption
        let layer3Result = try layer3Decrypt(data: encrypted, key: keys.key3, iv: iv)
        
        // Layer 2: Reverse bit manipulation
        let layer2Result = layer2Manipulate(data: layer3Result, key: keys.key2)
        
        // Layer 1: Reverse bitwise scrambling
        let layer1Result = layer1Scramble(data: layer2Result, key: keys.key1)
        
        // Remove padding
        let originalData = layer1Result.prefix(layer1Result.count - paddingLength)
        
        return String(data: originalData, encoding: .utf8)!
    }
    
    /// Generate a new key pair for asymmetric operations (if needed)
    public func generateKeyPair() throws -> (privateKey: String, publicKey: String) {
        var privateKey = Data(count: 32)
        _ = privateKey.withUnsafeMutableBytes { pointer in
            SecRandomCopyBytes(kSecRandomDefault, 32, pointer.baseAddress!)
        }
        
        let publicKey = hashPassword(privateKey.base64EncodedString())
        
        return (
            privateKey: privateKey.base64EncodedString(),
            publicKey: publicKey.base64EncodedString()
        )
    }
} 

File: languages/csharp/XANDEncrypt.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;

namespace XANDEncrypt
{
    /// <summary>
    /// Holds all the pieces we need to decrypt our data later
    /// Think of this like a secure envelope with all the instructions inside
    /// </summary>
    public class EncryptedData
    {
        public string Encrypted { get; set; } = string.Empty;
        public string Iv { get; set; } = string.Empty;
        public string Nonce { get; set; } = string.Empty;
        public long Timestamp { get; set; }
        public int PaddingLength { get; set; }

        /// <summary>
        /// Convert our encrypted data to a JSON string for easy storage
        /// </summary>
        public string ToJson()
        {
            return JsonSerializer.Serialize(this);
        }

        /// <summary>
        /// Create encrypted data from a JSON string
        /// </summary>
        public static EncryptedData FromJson(string json)
        {
            return JsonSerializer.Deserialize<EncryptedData>(json) ?? new EncryptedData();
        }
    }

    /// <summary>
    /// The main encryption class - this is what you'll use to protect your data
    /// It handles all the complex cryptography so you don't have to worry about it
    /// </summary>
    public class XANDEncrypt : IDisposable
    {
        // Our configuration constants
        private const int KeySize = 32;        // 256 bits
        private const int NonceSize = 16;      // 128 bits
        private const int PaddingSize = 16;    // Max padding
        private const int IvSize = 16;         // AES IV size

        // The password that protects all our data
        private readonly string _password;
        private bool _disposed = false;

        /// <summary>
        /// Create a new encryption system with your password - keep it safe!
        /// </summary>
        public XANDEncrypt(string password)
        {
            _password = password ?? throw new ArgumentNullException(nameof(password));
        }

        /// <summary>
        /// Clean up any sensitive data when we're done
        /// </summary>
        public void Dispose()
        {
            if (!_disposed)
            {
                // Clear any sensitive data from memory
                _disposed = true;
            }
        }

        /// <summary>
        /// Convert password to SHA-256 hash using .NET cryptography APIs
        /// SHA256.Create() uses native crypto libraries for optimized performance
        /// 'using' statement ensures proper disposal of cryptographic resources
        /// </summary>
        private byte[] HashPassword(string password)
        {
            using var sha256 = SHA256.Create();
            return sha256.ComputeHash(Encoding.UTF8.GetBytes(password));
        }

        /// <summary>
        /// Generate three time-dependent keys using HMAC-SHA256
        /// BitConverter.GetBytes() handles endianness conversion automatically
        /// Tuple return type provides multiple values without out parameters
        /// </summary>
        private (byte[] key1, byte[] key2, byte[] key3, long timestamp) GenerateTimeShiftedKeys(byte[] baseKey, long timestamp)
        {
            var timeBuffer = BitConverter.GetBytes(timestamp);
            if (BitConverter.IsLittleEndian)
                Array.Reverse(timeBuffer);

            // Each layer gets its own unique key derived from the previous one
            // C# HMAC is thread-safe and uses native crypto libraries when available
            var key1 = ComputeHmac(baseKey, timeBuffer);
            var key2 = ComputeHmac(key1, Encoding.UTF8.GetBytes("layer2"));
            var key3 = ComputeHmac(key2, Encoding.UTF8.GetBytes("layer3"));

            return (key1, key2, key3, timestamp);
        }

        /// <summary>
        /// Generate cryptographically secure random nonce and padding
        /// RandomNumberGenerator.Create() uses OS random number generator
        /// 'using' statement ensures proper disposal of RNG instance
        /// </summary>
        private (byte[] nonce, byte[] padding) GenerateNonceAndPadding()
        {
            var nonce = new byte[NonceSize];
            using (var rng = RandomNumberGenerator.Create())
            {
                rng.GetBytes(nonce);
            }

            var paddingLength = (nonce[0] % PaddingSize) + 1;
            var padding = new byte[paddingLength];
            using (var rng = RandomNumberGenerator.Create())
            {
                rng.GetBytes(padding);
            }

            return (nonce, padding);
        }

        /// <summary>
        /// First encryption layer: XOR and bit rotation operations
        /// C# bitwise operations operate on int, requiring byte casting for 8-bit values
        /// JIT compiler optimizes bitwise operations to efficient machine code
        /// </summary>
        private byte[] Layer1Scramble(byte[] data, byte[] key)
        {
            var result = new byte[data.Length];

            for (int i = 0; i < data.Length; i++)
            {
                var keyByte = key[i % key.Length];
                var dataByte = data[i];

                // Do some bit magic to scramble the data
                // C# bitwise ops work on int, so we cast back to byte after operations
                var scrambled = (byte)(dataByte ^ keyByte);
                scrambled = (byte)(((scrambled << 3) | (scrambled >> 5)) & 0xFF); // Rotate left by 3
                scrambled = (byte)(scrambled ^ (((keyByte << 1) | (keyByte >> 7)) & 0xFF));

                result[i] = scrambled;
            }

            return result;
        }

        /// <summary>
        /// Second encryption layer: Position-dependent bit rotation
        /// Rotation amount calculated from byte position and key value
        /// Conditional rotation prevents unnecessary operations when rotation is zero
        /// </summary>
        private byte[] Layer2Manipulate(byte[] data, byte[] key)
        {
            var result = new byte[data.Length];

            for (int i = 0; i < data.Length; i++)
            {
                var keyByte = key[i % key.Length];
                var dataByte = data[i];

                // Rotate bits based on position and key
                var rotation = (i + keyByte) % 8;
                var manipulated = dataByte;

                if (rotation > 0)
                {
                    manipulated = (byte)(((manipulated << rotation) | (manipulated >> (8 - rotation))) & 0xFF);
                }

                // Mix in some position-dependent randomness
                manipulated = (byte)(manipulated ^ ((keyByte + i) & 0xFF));

                result[i] = manipulated;
            }

            return result;
        }

        /// <summary>
        /// Third encryption layer: AES-256-CBC encryption
        /// Uses .NET Aes class with PKCS7 padding for standard compliance
        /// CBC mode requires unique IV for each encryption operation
        /// </summary>
        private (byte[] encrypted, byte[] iv) Layer3Encrypt(byte[] data, byte[] key)
        {
            using var aes = Aes.Create();
            aes.Key = key;
            aes.Mode = CipherMode.CBC;
            aes.Padding = PaddingMode.PKCS7;

            var iv = aes.IV;
            using var encryptor = aes.CreateEncryptor();
            var encrypted = encryptor.TransformFinalBlock(data, 0, data.Length);

            return (encrypted, iv);
        }

        /// <summary>
        /// Third layer decryption: Reverse the AES-256 encryption
        /// </summary>
        private byte[] Layer3Decrypt(byte[] data, byte[] key, byte[] iv)
        {
            using var aes = Aes.Create();
            aes.Key = key;
            aes.IV = iv;
            aes.Mode = CipherMode.CBC;
            aes.Padding = PaddingMode.PKCS7;

            using var decryptor = aes.CreateDecryptor();
            return decryptor.TransformFinalBlock(data, 0, data.Length);
        }

        /// <summary>
        /// Apply key decay by XORing with time-based hash
        /// Prevents key reuse attacks by modifying keys after each operation
        /// Uses current timestamp to ensure uniqueness of decay value
        /// </summary>
        private void ApplyKeyDecay(ref byte[] key1, ref byte[] key2, ref byte[] key3)
        {
            var timestamp = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
            var timeStr = timestamp.ToString();

            // Mix in some fresh randomness to age the keys
            var keys = new[] { key1, key2, key3 };
            for (int i = 0; i < keys.Length; i++)
            {
                var decayValue = ComputeHmac(keys[i], Encoding.UTF8.GetBytes(timeStr));
                for (int j = 0; j < keys[i].Length; j++)
                {
                    keys[i][j] ^= decayValue[j % decayValue.Length];
                }
            }

            key1 = keys[0];
            key2 = keys[1];
            key3 = keys[2];
        }

        /// <summary>
        /// Compute HMAC-SHA256 for key derivation
        /// </summary>
        private byte[] ComputeHmac(byte[] key, byte[] data)
        {
            using var hmac = new HMACSHA256(key);
            return hmac.ComputeHash(data);
        }

        /// <summary>
        /// Convert bytes to a hex string for easy storage
        /// </summary>
        private string BytesToHex(byte[] bytes)
        {
            return Convert.ToHexString(bytes).ToLower();
        }

        /// <summary>
        /// Convert a hex string back to bytes
        /// </summary>
        private byte[] HexToBytes(string hex)
        {
            return Convert.FromHexString(hex);
        }

        /// <summary>
        /// Main encryption function implementing triple-layer encryption
        /// Processes data through bit scrambling, position-dependent rotation, and AES-256
        /// Returns EncryptedData object containing encrypted data and metadata
        /// </summary>
        public EncryptedData Encrypt(string data)
        {
            if (string.IsNullOrEmpty(data))
                throw new ArgumentException("Data cannot be null or empty", nameof(data));

            // Generate our base key from the password
            var baseKey = HashPassword(_password);

            // Create three unique keys that change over time
            var timestamp = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
            var (key1, key2, key3) = GenerateTimeShiftedKeys(baseKey, timestamp);

            // Generate some random data to make each encryption unique
            var (nonce, padding) = GenerateNonceAndPadding();

            // Prepare our data with padding
            var dataBytes = Encoding.UTF8.GetBytes(data);
            var paddedData = new byte[dataBytes.Length + padding.Length];
            Array.Copy(dataBytes, paddedData, dataBytes.Length);
            Array.Copy(padding, 0, paddedData, dataBytes.Length, padding.Length);

            // First layer: Mix up the data with some fancy bit operations
            var layer1Result = Layer1Scramble(paddedData, key1);

            // Second layer: More bit manipulation, but this time it depends on position
            var layer2Result = Layer2Manipulate(layer1Result, key2);

            // Third layer: The heavy artillery - AES-256 encryption
            var (layer3Result, iv) = Layer3Encrypt(layer2Result, key3);

            // After we use the keys, we modify them slightly
            var key1Copy = key1.ToArray();
            var key2Copy = key2.ToArray();
            var key3Copy = key3.ToArray();
            ApplyKeyDecay(ref key1Copy, ref key2Copy, ref key3Copy);

            // Package everything up for storage
            return new EncryptedData
            {
                Encrypted = BytesToHex(layer3Result),
                Iv = BytesToHex(iv),
                Nonce = BytesToHex(nonce),
                Timestamp = timestamp,
                PaddingLength = padding.Length
            };
        }

        /// <summary>
        /// Decrypt data that was encrypted with this class
        /// This reverses all three layers to get your original data back
        /// </summary>
        public string Decrypt(EncryptedData encryptedData)
        {
            if (encryptedData == null)
                throw new ArgumentNullException(nameof(encryptedData));

            // Generate our base key from the password
            var baseKey = HashPassword(_password);

            // Recreate the three keys using the same timestamp
            var (key1, key2, key3) = GenerateTimeShiftedKeys(baseKey, encryptedData.Timestamp);

            // Convert hex strings back to bytes
            var encrypted = HexToBytes(encryptedData.Encrypted);
            var iv = HexToBytes(encryptedData.Iv);
            var nonce = HexToBytes(encryptedData.Nonce);

            // Third layer: Reverse the AES-256 encryption
            var layer3Result = Layer3Decrypt(encrypted, key3, iv);

            // Second layer: Reverse the bit manipulation
            var layer2Result = Layer2Manipulate(layer3Result, key2);

            // First layer: Reverse the bit scrambling
            var layer1Result = Layer1Scramble(layer2Result, key1);

            // Remove the padding to get our original data
            var originalLength = layer1Result.Length - encryptedData.PaddingLength;
            var originalData = new byte[originalLength];
            Array.Copy(layer1Result, originalData, originalLength);

            return Encoding.UTF8.GetString(originalData);
        }

        /// <summary>
        /// Generate a key pair for asymmetric operations
        /// This creates a public and private key for advanced use cases
        /// </summary>
        public (string privateKey, string publicKey) GenerateKeyPair()
        {
            var privateKeyBytes = new byte[KeySize];
            using (var rng = RandomNumberGenerator.Create())
            {
                rng.GetBytes(privateKeyBytes);
            }

            var publicKey = HashPassword(Convert.ToBase64String(privateKeyBytes));

            return (
                Convert.ToBase64String(privateKeyBytes),
                Convert.ToBase64String(publicKey)
            );
        }
    }
} 